lxd/auth/oidc/oidc.go

```diff
@@ -371,6 +375,51 @@ func (*Verifier) setCookies(w http.ResponseWriter, idToken string, refreshToken
 	return nil
 }
 
+// secureCookieFromSession returns a *securecookie.SecureCookie that is secure, unique to each client, and possible to
+// decrypt on all cluster members.
+//
+// To do this we use the cluster private key as an input seed to HKDF (https://datatracker.ietf.org/doc/html/rfc5869) and
+// use the given sessionID uuid.UUID as a salt. The session ID can then be stored as a plaintext cookie so that we can
+// regenerate the keys upon the next request.
+//
+// Warning: Changes to this function might cause all existing OIDC users to be logged out of LXD (but not logged out of
+// the IdP).
+func (o *Verifier) secureCookieFromSession(sessionID uuid.UUID) (*securecookie.SecureCookie, error) {
+	// Get the sessionID as a binary so that we can use it as a salt.
+	salt, err := sessionID.MarshalBinary()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to marshal session ID as binary: %w", err)
+	}
+
+	// Get the current cluster private key.
+	clusterPrivateKey := o.clusterCert().PrivateKey()
+
+	// Extract a pseudo-random key from the cluster private key.
+	prk := hkdf.Extract(sha512.New, clusterPrivateKey, salt)
+
+	// Add "info" context information and perform expansion.
+	keyDerivationFunc := hkdf.Expand(sha512.New, prk, []byte("INTEGRITY"))
+
+	// Read 64 bytes of the derived key. The securecookie library recommends 64 bytes for the hash key (https://github.com/gorilla/securecookie).
+	cookieHashKey := make([]byte, 64)
+	_, err = io.ReadFull(keyDerivationFunc, cookieHashKey)
+	if err != nil {
+		return nil, fmt.Errorf("Failed creating secure cookie hash key: %w", err)
+	}
+
+	// Add "info" context information and perform expansion.
+	keyDerivationFunc = hkdf.Expand(sha512.New, prk, []byte("ENCRYPTION"))
```

Comment:
Could you add a comment here explaining the use of ENCRYPTION ?

Reply:
I've added comments for both and linked to the info section in the RFC :)